clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -2397,7 +2399,15 @@ public ProducerRecord<byte[], byte[]> onSend(ProducerRecord<byte[], byte[]> reco
         }
 
         @Override
-        public void onAcknowledgement(RecordMetadata metadata, Exception exception) {
+        public void onAcknowledgement(RecordMetadata metadata, Exception exception, Headers headers) {
+            if (headers == null) {
+                return;
+            }
+            if (!(headers instanceof RecordHeaders)) {
+                return;
+            }
+            RecordHeaders recordHeaders = (RecordHeaders) headers;
```

Comment:
I can't see any test setting header as `RecordHeaders` so am not sure what the benefit of this assertion if there is no actual test for it.

Reply:
Well, I think it is important that they are read-only so this assertion probably does have some value.

Reply:
Yes, the goal is to verify it is read only

Turn out this code is not hit during test. It is reference in one test, but the test does not send any record. I will fix it.
https://github.com/apache/kafka/blob/6c582bd2c5644bdbe2b92f6a50547ed27ae3908a/clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java#L2061-L2075

Reply:
https://github.com/apache/kafka/commit/988c09f48135a4dc1ab7503cf6ed45ea115063fa